src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,449 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Jaroslaw Stelter <jaroslaw.stelter@intel.com>
+//         Pawel Dobrowolski<pawelx.dobrowolski@intel.com>
+
+/*
+ * Dynamic module loading functions.
+ */
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/component_ext.h>
+#include <sof/common.h>
+#include <sof/ipc/topology.h>
+#include <sof/lib/wait.h>
+
+#include <sof/sof.h>
+#include <sof/spinlock.h>
+#include <sof/lib_manager.h>
+#include <sof/audio/module_adapter/module/generic.h>
+
+#ifdef LIBRARY_MANAGER
+#include <sof/audio/module_adapter/module/intel_modules.h>
+#endif
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/* 54cf5598-8b29-11ec-a8a3-0242ac120002 */
+
+DECLARE_SOF_UUID("lib_manager", lib_manager_uuid, 0x54cf5598, 0x8b29, 0x11ec,
+		 0xa8, 0xa3, 0x02, 0x42, 0xac, 0x12, 0x00, 0x02);
+
+DECLARE_TR_CTX(lib_manager_tr, SOF_UUID(lib_manager_uuid), LOG_LEVEL_INFO);
+
+/* TODO: Manifest size should be defined somewhere - to check!!!*/
+#define CAVS18_MAX_MANIFEST_SIZE  (38*1024)
+#define DMA_ELEM_SIZE		32
+#define PAGE_ALIGN			__aligned(HOST_PAGE_SIZE)
+
+/**
+ * DMA buffer
+ */
+struct lib_manager_dma_buf {
+	uintptr_t w_ptr;	/**< write pointer */
+	uintptr_t r_ptr;	/**< read pointer */
+	uintptr_t addr;		/**< buffer start pointer */
+	uintptr_t end_addr;	/**< buffer end pointer */
+	uint32_t size;		/**< buffer size */
+	uint32_t avail;		/**< buffer avail data */
+};
+
+struct lib_manager_dma_ext {
+	struct dma_copy dma;
+	struct dma_chan_data *chan;
+	struct lib_manager_dma_buf dmabp;
+};
+
+struct ext_library loader_ext_lib;
+
+void *lib_manager_get_config(struct comp_dev *dev)
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct module_data *md = &mod->priv;
+	struct module_config *dst = &md->cfg;
+
+	return dst->data;
+}
+
+static uint8_t *lib_manager_get_instance_bss_address(uint32_t module_id, uint32_t instance_id,
```

Comment:
Do we have UT for any of these APIs ? It seems like these can be tested quite easily. No need immediately, but should be added at some point.

Reply:
A UT would create a test module and attempt to load it on host using these APIs. i.e. the UT locally mallocs buffers and uses these as fake text, data and bss. The UT then send this fake structure to the APIs and validates the fake data is processed correctly. Fwiw, the UTs are also run under the valgrind memory checker.